src/main/java/com/google/sps/servlets/ChallengesServlet.java

```diff
@@ -13,38 +13,130 @@
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.api.client.json.JsonFactory;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultList;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
 
-import com.google.gson.Gson;
 import com.google.sps.data.Challenge;
 import com.google.sps.data.ChallengeData;
+import com.google.sps.data.User;
+
 import java.io.IOException;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.Scanner;
+
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import javafx.util.Pair;
+
 /** Servlet that manages challenges **/
 @WebServlet("/challenges")
 public class ChallengesServlet extends HttpServlet {
+  private static final String CLIENT_ID = "605480199600-e4uo1livbvl58cup3qtd1miqas7vspcu.apps.googleusercontent.com";
+  private static final HttpTransport HTTP_TRANSPORT = new UrlFetchTransport();
+  private static final JsonFactory JSON_FACTORY = new GsonFactory();
+  private static final GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(HTTP_TRANSPORT, JSON_FACTORY)
+                       .setAudience(Collections.singletonList(CLIENT_ID))
+                       .build();
+
+  private static final int NO_CHALLENGES = 0;
+  private static final String NUM_CHALLENGES = "num-challenges";
+  private static final String ID_TOKEN = "id-token";
+  private ArrayList<Challenge> requested_challenge_list = new ArrayList<>();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int num_challenges = getNumChallenges(request);
+    GoogleIdToken id_token = verifyId(request.getParameter(ID_TOKEN));
+    if (id_token == null) {
+        System.out.println("Invalid ID token.");
+        response.setStatus(400);
+        return;
+    }
+  
+    Payload payload = id_token.getPayload();
+    String user_id = payload.getSubject();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query(User.DATA_TYPE).setFilter(new FilterPredicate(User.ID, FilterOperator.EQUAL, user_id));
+    Entity entity = datastore.prepare(query).asSingleEntity();
```

Comment:
I believe it already is

Reply:
Oh nice, in that case you could use datastore.get(key)

Reply:
i believe the entity key is private so I don't necessarily have access

Reply:
To expand on Erick's comment: in this case you want to lookup by the Google ID payload's getSubject() string, rather than datastore's primary key. Therefore, event if entity key is public, you wouldn't want to use `datastore.get(key)` to lookup the user.